### PR TITLE
Add http retry policy

### PR DESCRIPTION
**Pull request by madhukar-agsft**
1) Add Http Retry Policy for SimulationClient 
2) Create Test project
3) Add Uni Tests with urls `http://www.google.com` for success and `https://3dsimulationNoSiteExistsWiththisName.com` (a not reachable site) for negative scenario as implementing mock setup for HttpRetryMessageHandler and/or HttpClient did not work. It was not calling the handler's SendAsync method which has Retry policy ExecuteAsync.

### CODE REVIEW

**Review by alexDrinkwater — APPROVED** [review]

**Comment by madhukar-agsft** [comment]
run test

**Comment by pkrull** [comment]
run test

File: test/SimulationCSharpClient.Tests/HttpRetryHandler/HttpRetryMessageHandlerTests.cs [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+        [Test]
+        public void HttpGet_Success()
+        {
+            // Arrange
+            var retryMessageHandler = new HttpRetryMessageHandler(6);
+
+            using (var httpClient = new HttpClient(retryMessageHandler))
+            {
+                // Act
+                var response = httpClient.GetStringAsync("http://www.google.com").Result;
```
Comment (chadqueen) [inline test/SimulationCSharpClient.Tests/HttpRetryHandler/HttpRetryMessageHandlerTests.cs]: should not rely upon an internet connection in general for unit tests. would be better to mock the httpclient

File: test/SimulationCSharpClient.Tests/HttpRetryHandler/HttpRetryMessageHandlerTests.cs [hunk 2]
```diff
@@ -0,0 +1,82 @@
...
+        public void HttpGet_Fail_But_RetryForMaxTries()
+        {
+            int maxReTries = 3;
+            var httpRetryMessageHandler = new HttpRetryMessageHandler(maxReTries);
+            using (var httpClient = new HttpClient(httpRetryMessageHandler))
+            {
+                try
+                {
+                    // Act
+                    var response = httpClient.GetStringAsync("https://3dsimulationNoSiteExistsWiththisName.com").Result;
```
Comment (chadqueen) [inline test/SimulationCSharpClient.Tests/HttpRetryHandler/HttpRetryMessageHandlerTests.cs]: mock httpclient

File: test/SimulationCSharpClient.Tests/HttpRetryHandler/HttpRetryMessageHandlerTests.cs [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+        {
+            int maxReTries = -1;
+            int expectedMaxReTries = 6;
+            var httpRetryMessageHandler = new HttpRetryMessageHandler(maxReTries);
+            using (var httpClient = new HttpClient(httpRetryMessageHandler))
+            {
+                try
+                {
+                    // Act
+                    var response = httpClient.GetStringAsync("https://3dsimulationNoSiteExistsWiththisName.com").Result;
```
Comment (chadqueen) [inline test/SimulationCSharpClient.Tests/HttpRetryHandler/HttpRetryMessageHandlerTests.cs]: mock httpclient

File: src/SimulationCSharpClient/Client/SimulationClient.cs [hunk 4]
```diff
@@ -23,14 +23,16 @@ public partial class SimulationClient
         /// <param name="secret">client secret</param>
         /// <param name="audience">client audience</param>
         /// <param name="httpClient">injected http client</param>
-        public SimulationClient(string baseUrl, string auth0TokenURL, string client, string secret, string audience, HttpClient httpClient = null)
+        /// <param name="maxRetries">http maxRetries.</param>
+        /// <param name="handler">http client handler.</param>
+        public SimulationClient(string baseUrl, string auth0TokenURL, string client, string secret, string audience, HttpClient httpClient = null, int maxRetries = 6, HttpClientHandler handler = null)
```
Comment (chadqueen) [inline src/SimulationCSharpClient/Client/SimulationClient.cs:28]: does it need to receive in a HttpClientHandler? would most calls to this client pretty much always use the same/similar handler? i think at a minimum you can check if it is null and provide a default if it is not provided

**Review by chadqueen — CHANGES_REQUESTED** [review]

**Comment by pkrull** [comment]
I'd like the unit tests to complete in less than a minute. Can we adjust the timeout period and max retries to allow this and still effectively test it?

File: src/SimulationCSharpClient/Client/SimulationClient.cs (hunk 4 above)
Comment (madhukar-agsft) [reply src/SimulationCSharpClient/Client/SimulationClient.cs:28]: We are just passing this handler to HttpRetryMessageHandler and where it is creating **HttpClientHandler()** object if it is null

`public HttpRetryMessageHandler(int maxRetries = 6, HttpClientHandler handler = null) : base(handler ?? new HttpClientHandler())`

File: test/SimulationCSharpClient.Tests/HttpRetryHandler/HttpRetryMessageHandlerTests.cs (hunk 1 above)
Comment (madhukar-agsft) [reply test/SimulationCSharpClient.Tests/HttpRetryHandler/HttpRetryMessageHandlerTests.cs]: Resolved by creating TestHandlers for Success and Failures and with Dummy Apis

File: test/SimulationCSharpClient.Tests/HttpRetryHandler/HttpRetryMessageHandlerTests.cs (hunk 2 above)
Comment (madhukar-agsft) [reply test/SimulationCSharpClient.Tests/HttpRetryHandler/HttpRetryMessageHandlerTests.cs]: Resolved by creating TestHandlers for Success and Failures and with Dummy Apis

File: test/SimulationCSharpClient.Tests/HttpRetryHandler/HttpRetryMessageHandlerTests.cs (hunk 3 above)
Comment (madhukar-agsft) [reply test/SimulationCSharpClient.Tests/HttpRetryHandler/HttpRetryMessageHandlerTests.cs]: Resolved by creating TestHandlers for Success and Failures and with Dummy Apis

**Comment by madhukar-agsft** [comment]
> I'd like the unit tests to complete in less than a minute. Can we adjust the timeout period and max retries to allow this and still effectively test it?

Yes Reduced MaxRetries and  removed one test case which was to assert if Handler retires for default number of retries(6 retries ~3 min). Now it is taking around 8 secs to run tests

**Review by pkrull — APPROVED** [review]

**Review by chadqueen — APPROVED** [review]

**Pull request merged**